### PR TITLE
Beginnings of a class for converting different representations of a Dart source file

### PR DESCRIPTION
**Pull request by alan-knight**
Beginnings of a class for converting different representations of a Dart source file

### CODE REVIEW

File: dwds/lib/src/dart_uri.dart [hunk 1]
```diff
@@ -0,0 +1,42 @@
+import 'package:path/path.dart' as p;
```
Comment (jakemac53) [inline dwds/lib/src/dart_uri.dart]: copyright

File: dwds/lib/src/dart_uri.dart [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+import 'package:vm_service_lib/vm_service_lib.dart';
+
+// The URI for a particular Dart file, able to canonicalize from various
+/// different representations.
+// TODO: Support absolute file and http URLs and org-dartlang-app.
+class DartUri {
+  /// Expects a URI of the form /hello_world/main.dart or /packages/...
+  DartUri.fromSourcemap(String dartFile) {
+    // TODO: What's the exact form of /packages URLS? Do we need to
```
Comment (jakemac53) [inline dwds/lib/src/dart_uri.dart]: you can remove this todo, we are doing the right thing here :) (there will be no lib dir)

File: dwds/lib/src/dart_uri.dart [hunk 3]
```diff
@@ -0,0 +1,42 @@
+import 'package:path/path.dart' as p;
+import 'package:vm_service_lib/vm_service_lib.dart';
+
+// The URI for a particular Dart file, able to canonicalize from various
+/// different representations.
+// TODO: Support absolute file and http URLs and org-dartlang-app.
+class DartUri {
+  /// Expects a URI of the form /hello_world/main.dart or /packages/...
```
Comment (jakemac53) [inline dwds/lib/src/dart_uri.dart]: nit: mention that the `/hello_world/main.dart` style are urls are relative to the directory being served (not the package root)

File: dwds/lib/src/dart_uri.dart [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    // TODO: What's the exact form of /packages URLS? Do we need to
+    // compensate for the lib directory?
+    if (dartFile.startsWith('/packages/')) {
+      dartForm = 'package:${dartFile.substring("/packages/".length)}';
+    } else {
+      dartForm = _noLeadingSlash(dartFile);
+    }
+  }
+
+  /// Expects a ScriptRef which provides the absolute URI, plus
```
Comment (jakemac53) [inline dwds/lib/src/dart_uri.dart]: This comment doesn't really match up with the current state of things :)

File: dwds/lib/src/dart_uri.dart [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    // e.g. hello_world/main.dart. In the short term the ScriptRef just gives us
+    // main.dart, so work around it.
+    var relative = script.uri;
+    dartForm = _noLeadingSlash(p.join(mainUri, relative));
+  }
+
+  /// Make a path relative by removing the leading slash if present.
+  String _noLeadingSlash(String s) => s[0] == '/' ? s.substring(1) : s;
+
+  String get uri => dartForm;
```
Comment (jakemac53) [inline dwds/lib/src/dart_uri.dart]: Do we need this one? What is it supposed to represent?

File: dwds/lib/src/dart_uri.dart [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+  /// The canonical Dart form of the URI.
+  ///
+  /// This is a relative URI, which can be used to fetch the corresponding file
+  /// from the server. For example, 'hello_world/main.dart'
+  String dartForm;
```
Comment (jakemac53) [inline dwds/lib/src/dart_uri.dart]: nit: consider renaming to `serverPath` or something along those lines?

File: dwds/lib/src/dart_uri.dart (hunk 1 above)
Comment (alan-knight) [reply dwds/lib/src/dart_uri.dart]: Done.

File: dwds/lib/src/dart_uri.dart (hunk 2 above)
Comment (alan-knight) [reply dwds/lib/src/dart_uri.dart]: Done.

File: dwds/lib/src/dart_uri.dart (hunk 3 above)
Comment (alan-knight) [reply dwds/lib/src/dart_uri.dart]: Done.

File: dwds/lib/src/dart_uri.dart (hunk 1 above)
Comment (natebosch) [reply dwds/lib/src/dart_uri.dart]: Do we need a push? I still don't see it.

File: dwds/lib/src/dart_uri.dart (hunk 5 above)
Comment (alan-knight) [reply dwds/lib/src/dart_uri.dart]: Removed.

File: dwds/lib/src/dart_uri.dart (hunk 6 above)
Comment (alan-knight) [reply dwds/lib/src/dart_uri.dart]: Went with serverUri

File: dwds/lib/src/dart_uri.dart (hunk 4 above)
Comment (alan-knight) [reply dwds/lib/src/dart_uri.dart]: Thoroughly revised the class, this constructor is gone now.

**Comment by alan-knight** [comment]
Major revision. Also now depends on correct org-dartlang-app: URIs, which just landed.

File: dwds/lib/src/dart_uri.dart (hunk 1 above)
Comment (alan-knight) [reply dwds/lib/src/dart_uri.dart]: Did a major revision. Pushed now.

File: dwds/lib/src/dart_uri.dart [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+  ///  - package:packageName/pathUnderLib/file.dart
+  ///  - org-dartlang-app:///prefix/path/file.dart, where prefix is ignored.
+  ///    e.g. org-dartlang-app:example/hello_world/main.dart,
+  ///  - /packages/packageName/foo.dart, the web server form of a package URI,
+  ///    e.g. /packages/path/src/utils.dart
+  ///  - /path/foo.dart, e.g. hello_world/web/main.dart, where path is a web
+  ///    server path and so relative to the directory being served, not to the
+  ///    package.
+  ///
+  /// Note that there was a bug where older SDKs might provide
```
Comment (jakemac53) [inline dwds/lib/src/dart_uri.dart]: nit: incomplete thought :D

File: dwds/lib/src/dart_uri.dart [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+    // TODO: To be able to convert to an org-dartlang-app: URI we will
+    // need to know the root - possibly keep it as a static?
+    serverUri = Uri.parse(uri).pathSegments.skip(1).join('/').toString();
+  }
+
+  /// Construct from a path of the form /packages/packageName/foo.dart
+  DartUri._fromPackagesPath(this.serverUri);
+
+  /// Construct from an ordinary path, relative to the directory being served.
+  DartUri._fromPath(String uri) {
```
Comment (jakemac53) [inline dwds/lib/src/dart_uri.dart]: nit: fromServePath?

File: dwds/lib/src/dart_uri.dart [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+    serverUri = uri[0] == '/' ? uri.substring(1) : uri;
+  }
+
+  /// The canonical web server form of the URI.
+  ///
+  /// This is a relative URI, which can be used to fetch the corresponding file
+  /// from the server. For example, 'hello_world/main.dart' or
+  /// 'packages/path/src/utils.dart'.
+  String serverUri;
```
Comment (jakemac53) [inline dwds/lib/src/dart_uri.dart]: nit: move to the top of the class, and make final (may need to make some other things be factories)

**Comment by jakemac53** [comment]
Oh, I didn't see this already got merged but at least the last comment (making the field final) should be done in a followup if you don't mind @alan-knight

File: dwds/lib/src/dart_uri.dart (hunk 7 above)
Comment (alan-knight) [reply dwds/lib/src/dart_uri.dart]: Done in #396

File: dwds/lib/src/dart_uri.dart (hunk 9 above)
Comment (alan-knight) [reply dwds/lib/src/dart_uri.dart]: Done in #396

File: dwds/lib/src/dart_uri.dart (hunk 8 above)
Comment (alan-knight) [reply dwds/lib/src/dart_uri.dart]: Done in #396

**Pull request merged**